### PR TITLE
Support for PixRacer board

### PR DESCRIPTION
**Pull request by digitalentity**
The board boots, gyro/acc/mag are detected.
image

What to check/do:
- [x] RCC
- [x] Gyroscope/Accelerometer
  - [x] MPU9250
  - [x] ICM-20608
  - [x] ICM-20608 DRDY interrupt
- [x] MS5611 baro (SPI bus)
- [ ] Compass
  - [x] MPU9250 built-in compass
  - [ ] HMC5983 compass (SPI bus)
- [x] USB
- [x] SPI bus
- [ ] I2C bus
- [ ] FM25V02-G FRAM memory
- [ ] UART7 and UART8
- [ ] SDCard
  - [ ] SDIO bus
- [x] LEDs
- [ ] RC input
  - [x] PPM
  - [ ] SerialRX
- [x] Motor output
- [x] Buzzer
- [ ] VBAT/Curr sensor
- [x] Bootloader/DFU

Link to schematic: https://pixhawk.org/_media/modules/pixracer-r14.pdf

### CODE REVIEW

**Comment by digitalentity** [comment]
@sambas did you manage to get DFU working on F4 (ANYFC)?

**Comment by sambas** [comment]
Yes, jump to bootloader and direct dfu flash works when using zadig installed driver on Windows. I think full erase doesnt work. I have uses colibri_race target in configurator to flash.

**Comment by digitalentity** [comment]
Hmm... for some reason the PixRacer doesn't jump to bootloader...

**Comment by sambas** [comment]
Check if bootloader start address is different in 427 and the ram address where it writes the magic word.

**Comment by digitalentity** [comment]
No luck so far, DFU doesn't get executed for some reason. Need to setup a full-blown debugger...

**Comment by digitalentity** [comment]
On the other hand, verified that PPM RX and motor outputs work so the board should probably be flyable.

**Comment by Maxzor** [comment]
Flyable! I would test it right now! Cannot flash the hex via GUI through DFU... is either stuck on verification, or programming fails. And cannot manage to setup SWD damn!

Edit : how could you setup a full debugger when there are only SWDIO and SWCK?

File: src/main/target/PIXRACER/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# AnyFC
```
Comment (DzikuVx) [inline src/main/target/PIXRACER/README.md]: Pixracer?

**Comment by digitalentity** [comment]
Hm... This is the first target with dual acc/gyro sensors, both from Invensence (`accgyro_mpu` driver). The problem is gyro interrupt request - current MPU driver only allows one IRQ line, while on PixRacer we need two.

Does anybody have any suggections how to make dual gyro interrupt lines happen?

We can't simply enable both because if before soft-reset we used one gyro and after soft-reset we'll be using a different one - we'll have DRDY interrupt enabled on both and will be received gyro interrupts from both gyros while configuring only one.

**Comment by digitalentity** [comment]
For now I've separated the PIXRACER target into two build flavours - PIXRACER (with MPU9250 gyro) and PIXRACER_ICM20608 (with ICM20608 gyro)

**Comment by oleost** [comment]
Does pixracer fly and work now?

All uarts working?
Is there invertion on uarts?

**Comment by digitalentity** [comment]
It's far from being fully functional. Not all UARTs are working (lacking UART7/8). Also, as in all F4 boards - no hardware inversion built in (only as a discrete inverter on certain UARTs).

**Comment by martinbudden** [comment]
@digitalentity , when you have a bit of time, can you rebase this?

**Comment by digitalentity** [comment]
@martinbudden I'd like to finish a few minor things - I2C, UART7 and UART8, SerialRX support, VBAT & current sensor support. Shouldn't take more than a few days depending on how much time I'll have to invest into this.

EDIT: Given it a second though - I'll merge it as is and fix things on the fly to avoid making big and messy PR which is a pain in the rear to rebase.

**Comment by digitalentity** [comment]
Target boots, gyro/accel seems to work. 
I managed to burn out a 3.3V LDO on my PixRacer board due to weird pinout of SWD connector - pin1 on 3.3V, not GND as on other boards :cry:. Now I have to find a new one and a fine soldering tip to replace it.

**Comment by martinbudden** [comment]
Yes, I'd merge this in as it is. It doesn't affect other targets and merging it now will avoid error-prone rebases in future.

**Comment by digitalentity** [comment]
Merged.

**Comment by olibye** [comment]
What's the view on when SerialRX will be supported? Any pointers on code I could copy/tweak from another F4 board?

**Comment by olibye** [comment]
I made the following commit. I'll check it flies and read the contribution guidelines before making a PR.

https://github.com/olibye/inav/commit/2f51ea9f89cbfa0f82a981c96ef373571d110ce7

**Comment by greybush** [comment]
is the pixracer and inav operational yet ? also is dshot going to be add to it ? i am very interested in trying this out

**Comment by olibye** [comment]
It flies, features as above, now include SerialRX (i.e. FrySky SBUS).
I'm now playing with s.port for telemetry back to to Taranis Radio.

DShot is iNav question, not specific to the PixRacer target so I'm not qualified to answer, but.... AFAIK the rumoured plan is to merge these code bases. I.e. iNav/Betaflight would share much of that stack?

**Comment by digitalentity** [comment]
I don't see much benefit in DShot (if any at all). Some boards might eventually get it supported, but only if it won't conflict with all modes of operation (airplane etc).

Pixracer is generally supported - not all of the hardware, but it's very much flyable.

**Comment by greybush** [comment]
is there any info on flashing process needed to run  inav/betaflight/raceflight on the pixracer  .hex files? i couldn't find any or maybe i wasn't looking in the right places

**Comment by greybush** [comment]
i look forward to the merger

**Comment by olibye** [comment]
Good point. You have to short the boot pads ([see schematic](https://pixhawk.org/_detail/modules/pixracer_r09_top_pinouts.png?id=modules%3Apixracer))  before power up (the board will show a solid green light) then you can load the firmware in DFU mode from the iNav configurator by selecting the PIXRACER board.

If you don't short the boot pads, the board will boot PX4 stack (fast flashing, then normal PX4 lights). The iNav configurator will can't put the device into DFU mode from there.

Once the iNav stack has loaded you may need to short the boot pads if verify fails ~50% of the time.

I could add this as a target readme or something

**Comment by digitalentity** [comment]
@olibye if you have the time please submit a PR to add flashing procedure to the docs. Thanks!

**Comment by anbello** [comment]
If I flash my pixracer with inav following info by @olibye then to return to arducopter do I have to flash px4 bootloader? Is there anyone that has gone both way without problems?
If is all OK I can do some testing for this target.
Thanks.

**Comment by olibye** [comment]
Yes, I've gone back to PX4 a few times. Use dfu-util to restore the NuttX boot loader.

`dfu-util -a 0 -d 0483:df11 --dfuse-address 0x08000000 -D ~/gits/Bootloader/px4fmuv4_bl.bin`

**Comment by olibye** [comment]
**WARNING** Make sure you **fly** and **PID tune** in iNAv's Acro Mode first.

iNav's auto level modes will not work without stable PIDs in Acro mode.

**Comment by anbello** [comment]
@olibye thanks for the info, where did you get the bin for px4 BL?

Is your warning peculiar to pixracer target or general? I already have tuned PID for my quad when I used it with inav and airbotF4, now I only changed FC with pixracer, I think using those PID is OK.

**Comment by olibye** [comment]
PX4 Firmware built from https://github.com/PX4/Bootloader

With respect to tuning PIDs if you already have good PIDs that's a start.

My comment was a general comment to for those people (I used to be one) who take off, fly and land in horizon centric modes. I had not really understood (although its obvious now) that these auto stabilised modes build on top of a stable acro/rate/model centric tunes.

E.g. Being able to instantly disarm on landing was critical, as simply contact with the ground (as it normal with landing!) caused me wild oscillations when tuning (forget the PX4 stick down and left/right strategy).

Honestly it *is* worth the effort. Acro/Air mode in Beta/Clean/iNav felt much more dialled in than PX4.

**Comment by anbello** [comment]
I put the pixracer in DFU mode but when I try to flash the firmware with inav configurator I obtain this on the log:
```
2017-08-11 @ 10:20:03 -- Running - OS: Linux, Chrome: 60.0.3112.78, Configurator: 1.7.2
2017-08-11 @ 10:20:17 -- USB device successfully opened with ID: 1
2017-08-11 @ 10:20:17 -- Detected device with total flash size 2048 kiB
2017-08-11 @ 10:20:18 -- USB device successfully closed
```
in the progress bar:
`Erasing ... `
but nothing happen.

**Comment by DzikuVx** [comment]
@anbello do you have `full chip erase` enabled? if so, disable it

**Comment by anbello** [comment]
@DzikuVx I tried both way, same result

**Comment by stronnag** [comment]
Tried using dfu-util from the shell? Some clues in https://transfer.sh/R9o0v/flash.sh, e.g.:
````
$ ./flash.sh obj/inav_1.7.3_PIXRACER.hex
````

**Comment by anbello** [comment]
I tried, it search /dev/ttyUSB0 or /dev/ttyACM0 but when in DFU mode you don't have a tty device.

**Comment by stronnag** [comment]
Can you please provide us with some actual data to help diagnose the issue:

* What is the output of running the script with the device in normal mode?
* In normal mode, what are the permissions in /dev/ttyACM0 ?
* In DFU mode, what is the output of `dfu-util -l` ?
* What udev rules do you have set up to manage device access in normal and DFU modes ?

**Comment by anbello** [comment]
I am sorry, now I have reassembled my quad with pixracer and cannot do other test, I will do it in the future.

Anyway from what I remember:

- What is the output of running the script with the device in normal mode?

I didn't run it in normal mode, I thought it was not the right mode, I was wrong?

- In normal mode, what are the permissions in /dev/ttyACM0 ?

root:dialout
my user is in dialout group, I use this daily without problems with AMP Planner and Pixracer and with iNav configurator and AirbotF4.

- In DFU mode, what is the output of dfu-util -l ?

I didn't try this, anyway I used dfu-util several times to flash bin of night_ghost fork of Ardupilot on my AirbotF4 without problems

- What udev rules do you have set up to manage device access in normal and DFU modes ?

SUBSYSTEM=="usb", ATTRS{idVendor}=="0483", ATTRS{idProduct}=="df11", MODE="0664", GROUP="plugdev"

my user is also in the plugdev group

**Comment by stronnag** [comment]
Thanks, that all looks fine. Pretty much matches my Arch setup. It will be interesting to see what `dfu-util -l` has to say when you have the opportunity.

**Comment by anbello** [comment]
```
andrea@galileo:~$ dfu-util -l
dfu-util 0.8

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2014 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to dfu-util@lists.gnumonks.org

Found DFU: [0483:df11] ver=2200, devnum=2, cfg=1, intf=0, alt=3, name="@Device Feature/0xFFFF0000/01*004 e", serial="3179375C3036"
Found DFU: [0483:df11] ver=2200, devnum=2, cfg=1, intf=0, alt=2, name="@OTP Memory /0x1FFF7800/01*512 e,01*016 e", serial="3179375C3036"
Found DFU: [0483:df11] ver=2200, devnum=2, cfg=1, intf=0, alt=1, name="@Option Bytes  /0x1FFFC000/01*016 e/0x1FFEC000/01*016 e", serial="3179375C3036"
Found DFU: [0483:df11] ver=2200, devnum=2, cfg=1, intf=0, alt=0, name="@Internal Flash  /0x08000000/04*016Kg,01*064Kg,07*128Kg,04*016Kg,01*064Kg,07*128Kg", serial="3179375C3036"
```
Ciao
Andrea

**Comment by stronnag** [comment]
That looks very normal, c.f. my working / always flashed with my script MATEK F405 AIO board.
````
Found DFU: [0483:df11] ver=2200, devnum=93, cfg=1, intf=0, path="1-5.4", alt=3, name="@Device Feature/0xFFFF0000/01*004 e", serial="3077395E3136"
Found DFU: [0483:df11] ver=2200, devnum=93, cfg=1, intf=0, path="1-5.4", alt=2, name="@OTP Memory /0x1FFF7800/01*512 e,01*016 e", serial="3077395E3136"
Found DFU: [0483:df11] ver=2200, devnum=93, cfg=1, intf=0, path="1-5.4", alt=1, name="@Option Bytes  /0x1FFFC000/01*016 e", serial="3077395E3136"
Found DFU: [0483:df11] ver=2200, devnum=93, cfg=1, intf=0, path="1-5.4", alt=0, name="@Internal Flash  /0x08000000/04*016Kg,01*064Kg,07*128Kg", serial="3077395E3136"
````

**Comment by olibye** [comment]
@digitalentity What would you like me to do so that this target is automatically built for each new version of iNav? It only made it into iNav v1.5 and I'd like to simply flash the new versions instead of having to build them ;-) Should I look at porting it upstream into BetaFlight?

Is there a better issue for this discussion? As this issue is essentially closed (merged).

**Comment by digitalentity** [comment]
@olibye there are a few things that should be made working - see top of this thread https://github.com/iNavFlight/inav/pull/466#issue-171180292

I don't have much time to invest into making them happen, but if somebody looks into it - I'll happily include PixRacer into releases.

**Comment by JiviteshDebata** [comment]
New  Schematic Link (Closest To R14) :  https://github.com/PX4/Hardware/files/93463/PIXRACER-R13.pdf
Changes Mentioned in : https://github.com/PX4/Hardware/files/93465/Pixracer_RC13_CEO_report.pdf
Final R14 Schematic : https://pixhawk.org/_media/modules/pixracer-r14.pdf
Original Thread : https://github.com/PX4/Hardware/issues/47

**Pull request merged**